File: 0394-decode-string/0394-decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        string ans = "";
        int curr = 0;
        // cout<<s<<endl;
        while(curr<s.size()){
            if(s[curr] >= '0' && s[curr] <= '9'){
                int end = curr;
                while(s[end] != '[') end++;
                int freq = stoi(s.substr(curr, end - curr));
                end++;
                curr = end;
                // cout<<" freq: "<<freq<<" curr: "<<curr;
                int count = 1;
                while(count>0){
                    if(s[end] == ']') count--;
                    else if(s[end] == '[') count++;
                    end++;
                }
                string tok = decodeString(s.substr(curr, end - curr - 1));
                for(int i = 0; i<freq; i++) ans += tok;
                curr = end;
            }
            else{
                int end = curr;
                while(end < s.size() && !(s[end] >= '0' && s[end] <= '9')) end++;
                ans += s.substr(curr, end - curr);
                curr = end;
            }
        }
        return ans;
    }
};

File: 0394-decode-string/0394-decode-string.go
func isDigit(c byte) bool {
    if c >= '0' && c <= '9' {
        return true
    } else {
        return false
    }
}
func isBase(s string) bool {
    for i := range s {
        if isDigit(s[i]) {
            return false
        }
    }
    return true
}
func decodeString(s string) string {
    // unicode.IsDigit()
    if isBase(s) {
        return s
    }
    ans := ""
    for i:=0; i<len(s); i++ {
        if isDigit(s[i]) {
            j := i + 1
            for isDigit(s[j]) {
                j++
            }
            freq, _ := strconv.ParseInt(s[i : j], 10, 0)
            i = j + 1
            curr, brack := 0, 1
            for brack != 0 {
                curr++
                if s[i + curr] == '[' {
                    brack++
                } else if s[i + curr] == ']' {
                    brack--
                }
            }
            sub := s[i : i + curr]
            i = i + curr
            sub = decodeString(sub)
            substr := ""
            var k int64
            for k = 0; k < freq; k++ {
                substr += sub
            }
            ans += substr
        } else {
            ans += string(s[i])
        }
    }
    return ans
}